File: components/GameControls.cpp
#include "GameControls.h"
#include "../style/StyleProps.h"

#include <QGridLayout>
#include <QHBoxLayout>
#include <QVBoxLayout>
#include <QPushButton>
#include <QWidget>
#include <QTimer>
#include <QAction>
#include <QKeySequence>
#include <QApplication>

GameControls::GameControls(QWidget* parent) : QWidget(parent) {
  setMinimumWidth(300);
  possessionTeam_ = "Home";
  buildUi();
  wireSignals();
  buildKeyboardShortcuts();
  hideFollowUpButtons();
}

void GameControls::setActiveMainButton(QPushButton* button) {
  if (activeMainButton_ == button) return;
  if (activeMainButton_) Style::setState(activeMainButton_, "activeMain", false);
  activeMainButton_ = button;
  if (activeMainButton_) Style::setState(activeMainButton_, "activeMain", true);
}

void GameControls::clearActiveMainButton() {
  if (!activeMainButton_) return;
  Style::setState(activeMainButton_, "activeMain", false);
  activeMainButton_ = nullptr;
}

void GameControls::buildUi() {
  auto* mainLayout = new QVBoxLayout(this);
  mainLayout->setContentsMargins(0, 0, 0, 0);
  mainLayout->setSpacing(12);

  // Main grid: row 0 = Turnover (full width), rows 1–3 = 3x3 event buttons
  auto* mainGridWidget = new QWidget(this);
  mainGridLayout_ = new QGridLayout(mainGridWidget);
  mainGridLayout_->setContentsMargins(0, 0, 0, 0);
  mainGridLayout_->setSpacing(4);

  turnoverButton_ = new QPushButton("Turnover — Possession: Home", mainGridWidget);
  turnoverButton_->setToolTip("Tab  Toggle possession (Home ↔ Away)");
  Style::setSize(turnoverButton_, "md");
  Style::setVariant(turnoverButton_, "gameControl");
  turnoverButton_->setFocusPolicy(Qt::NoFocus);
  turnoverButton_->setMinimumHeight(40);
  mainGridLayout_->addWidget(turnoverButton_, 0, 0, 1, 3);

  // Create main event buttons
  hit16ydButton_ = new QPushButton("16-yd play", mainGridWidget);
  hit50ydButton_ = new QPushButton("50-yd play", mainGridWidget);
  hit75ydButton_ = new QPushButton("75-yd play", mainGridWidget);
  enterDButton_ = new QPushButton("Circle Entry", mainGridWidget);
  shotButton_ = new QPushButton("Shot", mainGridWidget);
  goalButton_ = new QPushButton("Goal", mainGridWidget);
  pcButton_ = new QPushButton("PC", mainGridWidget);
  psButton_ = new QPushButton("PS", mainGridWidget);
  cardButton_ = new QPushButton("Card", mainGridWidget);

  QList<QPushButton*> mainButtons = {
    hit16ydButton_, hit50ydButton_, hit75ydButton_,
    enterDButton_, shotButton_, goalButton_,
    pcButton_, psButton_, cardButton_
  };

  for (auto* button : mainButtons) {
    Style::setSize(button, "md");
    Style::setVariant(button, "gameControl");
    button->setFocusPolicy(Qt::NoFocus);
    button->setMinimumHeight(40);
  }

  // Row 1: 16-yd play, 50-yd play, 75-yd play
  mainGridLayout_->addWidget(hit16ydButton_, 1, 0);
  mainGridLayout_->addWidget(hit50ydButton_, 1, 1);
  mainGridLayout_->addWidget(hit75ydButton_, 1, 2);
  // Row 2: Circle Entry, Shot, Goal
  mainGridLayout_->addWidget(enterDButton_, 2, 0);
  mainGridLayout_->addWidget(shotButton_, 2, 1);
  mainGridLayout_->addWidget(goalButton_, 2, 2);
  // Row 3: PC, PS, Card
  mainGridLayout_->addWidget(pcButton_, 3, 0);
  mainGridLayout_->addWidget(psButton_, 3, 1);
  mainGridLayout_->addWidget(cardButton_, 3, 2);

  // Follow-up buttons container (initially hidden)
  followUpContainer_ = new QWidget(this);
  followUpLayout_ = new QHBoxLayout(followUpContainer_);
  followUpLayout_->setContentsMargins(0, 0, 0, 0);
  followUpLayout_->setSpacing(8);

  mainLayout->addWidget(mainGridWidget);
  mainLayout->addWidget(followUpContainer_);
  mainLayout->addStretch(1);
}

void GameControls::wireSignals() {
  connect(turnoverButton_, &QPushButton::clicked, this, &GameControls::onTurnoverClicked);

  auto connectMain = [this](QPushButton* btn) {
    connect(btn, &QPushButton::clicked, this, [this, btn]() { flashButtonBorder(btn); });
    connect(btn, &QPushButton::clicked, this, &GameControls::onMainButtonClicked);
  };
  connectMain(hit16ydButton_);
  connectMain(hit50ydButton_);
  connectMain(hit75ydButton_);
  connectMain(enterDButton_);
  connectMain(shotButton_);
  connectMain(goalButton_);
  connectMain(pcButton_);
  connectMain(psButton_);
  connectMain(cardButton_);
}

void GameControls::onTurnoverClicked() {
  possessionTeam_ = (possessionTeam_ == "Home") ? "Away" : "Home";
  turnoverButton_->setText(QString("Turnover — Possession: %1").arg(possessionTeam_));
  emit possessionChanged(possessionTeam_);
}

void GameControls::buildKeyboardShortcuts() {
  Q_ASSERT(QApplication::instance() != nullptr);

  auto makeAction = [this](int key, auto handler) -> QAction* {
    auto* act = new QAction(this);
    act->setShortcut(QKeySequence(key));
    act->setShortcutContext(Qt::ApplicationShortcut);
    connect(act, &QAction::triggered, this, handler);
    this->addAction(act);
    return act;
  };

  // Tab: turnover (toggle possession Home/Away)
  makeAction(Qt::Key_Tab, [this]() {
    if (turnoverButton_ && turnoverButton_->isVisible() && turnoverButton_->isEnabled())
      turnoverButton_->click();
  });

  // Main grid 3x3 left-hand QWERTY: Q W E | A S D | Z X C
  hit16ydAction_ = makeAction(Qt::Key_Q, [this]() {
    if (hit16ydButton_ && hit16ydButton_->isVisible() && hit16ydButton_->isEnabled()) hit16ydButton_->click();
  });
  hit50ydAction_ = makeAction(Qt::Key_W, [this]() {
    if (hit50ydButton_ && hit50ydButton_->isVisible() && hit50ydButton_->isEnabled()) hit50ydButton_->click();
  });
  hit75ydAction_ = makeAction(Qt::Key_E, [this]() {
    if (hit75ydButton_ && hit75ydButton_->isVisible() && hit75ydButton_->isEnabled()) hit75ydButton_->click();
  });
  enterDAction_ = makeAction(Qt::Key_A, [this]() {
    if (enterDButton_ && enterDButton_->isVisible() && enterDButton_->isEnabled()) enterDButton_->click();
  });
  shotAction_ = makeAction(Qt::Key_S, [this]() {
    if (shotButton_ && shotButton_->isVisible() && shotButton_->isEnabled()) shotButton_->click();
  });
  goalAction_ = makeAction(Qt::Key_D, [this]() {
    if (goalButton_ && goalButton_->isVisible() && goalButton_->isEnabled()) goalButton_->click();
  });
  pcAction_ = makeAction(Qt::Key_Z, [this]() {
    if (pcButton_ && pcButton_->isVisible() && pcButton_->isEnabled()) pcButton_->click();
  });
  psAction_ = makeAction(Qt::Key_X, [this]() {
    if (psButton_ && psButton_->isVisible() && psButton_->isEnabled()) psButton_->click();
  });
  cardAction_ = makeAction(Qt::Key_C, [this]() {
    if (cardButton_ && cardButton_->isVisible() && cardButton_->isEnabled()) cardButton_->click();
  });

  // Follow-ups: map visible follow-up buttons to 1..9
  followUpNumberActions_.clear();
  for (int i = 1; i <= 9; ++i) {
    const int key = (i == 1) ? Qt::Key_1
                  : (i == 2) ? Qt::Key_2
                  : (i == 3) ? Qt::Key_3
                  : (i == 4) ? Qt::Key_4
                  : (i == 5) ? Qt::Key_5
                  : (i == 6) ? Qt::Key_6
                  : (i == 7) ? Qt::Key_7
                  : (i == 8) ? Qt::Key_8
                             : Qt::Key_9;

    auto* act = makeAction(key, [this, i]() {
      if (!followUpContainer_ || !followUpContainer_->isVisible()) return;
      const int idx = i - 1;
      if (idx < 0 || idx >= followUpButtons_.size()) return;
      auto* btn = followUpButtons_.at(idx);
      if (!btn || !btn->isVisible() || !btn->isEnabled()) return;
      btn->click();
    });
    followUpNumberActions_.append(act);
  }

  // Escape: discard follow-ups and save with empty follow-up
  escapeAction_ = makeAction(Qt::Key_Escape, [this]() {
    if (followUpStage_ == FollowUpStage::None || currentMainEvent_.isEmpty()) return;
    
    // Emit with empty follow-up
    emit gameEventMarked(currentMainEvent_, QString());
    clearActiveMainButton();
    hideFollowUpButtons();
    currentMainEvent_.clear();
    currentFirstFollowUp_.clear();
    followUpStage_ = FollowUpStage::None;
  });
}

void GameControls::onMainButtonClicked() {
  auto* button = qobject_cast<QPushButton*>(sender());
  if (!button) return;

  QString eventName = button->text();
  currentMainEvent_ = eventName;
  currentFirstFollowUp_.clear();
  followUpStage_ = FollowUpStage::None;
  setActiveMainButton(button);
  emit mainEventPressed(eventName);
  showFirstLevelFollowUps(eventName);
}

void GameControls::onFollowUpButtonClicked() {
  auto* button = qobject_cast<QPushButton*>(sender());
  if (!button) return;

  QString followUpName = button->text();
  if (followUpStage_ == FollowUpStage::FirstLevel) {
    currentFirstFollowUp_ = followUpName;

    const QStringList second = getSecondLevelFollowUps(currentMainEvent_, currentFirstFollowUp_);
    if (second.isEmpty()) {
      emit gameEventMarked(currentMainEvent_, currentFirstFollowUp_);
      clearActiveMainButton();
      hideFollowUpButtons();
      currentMainEvent_.clear();
      currentFirstFollowUp_.clear();
      followUpStage_ = FollowUpStage::None;
      return;
    }

    showSecondLevelFollowUps(currentMainEvent_, currentFirstFollowUp_);
    return;
  }

  if (followUpStage_ == FollowUpStage::SecondLevel) {
    const QString combined = currentFirstFollowUp_.isEmpty()
      ? followUpName
      : (currentFirstFollowUp_ + " → " + followUpName);

    emit gameEventMarked(currentMainEvent_, combined);
    clearActiveMainButton();
    hideFollowUpButtons();
    currentMainEvent_.clear();
    currentFirstFollowUp_.clear();
    followUpStage_ = FollowUpStage::None;
    return;
  }

  // Fallback: treat as first-level
  emit gameEventMarked(currentMainEvent_, followUpName);
  clearActiveMainButton();
  hideFollowUpButtons();
  currentMainEvent_.clear();
  currentFirstFollowUp_.clear();
  followUpStage_ = FollowUpStage::None;
}

QStringList GameControls::getFirstLevelFollowUps(const QString& mainEvent) const {
  if (mainEvent == "Shot") {
    return {"On target", "Off target"};
  }
  if (mainEvent == "Circle Entry") {
    return {"Dribling", "Pass", "Deflection"};
  }
  if (mainEvent == "PC") {
    return {"Direct shot", "Variant", "Ruined"};
  }
  if (mainEvent == "75-yd play") {
    return {"Forward", "Sideways", "Back"};
  }
  if (mainEvent == "Card") {
    return {"Green", "Yellow", "Red"};
  }
  return {};
}

QStringList GameControls::getSecondLevelFollowUps(const QString& mainEvent, const QString& firstFollowUp) const {
  if (mainEvent == "Shot") {
    if (firstFollowUp == "On target") return {"Goal", "Saved", "Post"};
    if (firstFollowUp == "Off target") return {"Closeby", "Not close"};
    return {};
  }

  if (mainEvent == "PC") {
    if (firstFollowUp == "Direct shot") return {"Hit", "Swept", "Dragflick"};
    return {};
  }

  if (mainEvent == "Circle Entry") {
    // Second-level follow-ups apply to all first-level options (Dribling/Pass/Deflection)
    return {"Left", "Middle", "Right"};
  }

  return {};
}

void GameControls::showFirstLevelFollowUps(const QString& mainEvent) {
  // Clear existing follow-up buttons
  hideFollowUpButtons();

  QStringList actions = getFirstLevelFollowUps(mainEvent);
  
  // If no follow-up actions, emit the main event directly and return
  if (actions.isEmpty()) {
    emit gameEventMarked(mainEvent);
    clearActiveMainButton();
    currentMainEvent_.clear();
    currentFirstFollowUp_.clear();
    followUpStage_ = FollowUpStage::None;
    return;
  }

  // Create new follow-up buttons
  for (const QString& action : actions) {
    auto* button = new QPushButton(action, followUpContainer_);
    Style::setSize(button, "md");
    Style::setVariant(button, "gameControlFollowUp");
    button->setFocusPolicy(Qt::NoFocus);
    button->setMinimumHeight(40);
    
    // Connect click: flash first, then handle
    connect(button, &QPushButton::clicked, this, [this, button]() { flashButtonBorder(button); });
    connect(button, &QPushButton::clicked, this, &GameControls::onFollowUpButtonClicked);
    
    followUpLayout_->addWidget(button);
    followUpButtons_.append(button);
  }

  followUpStage_ = FollowUpStage::FirstLevel;
  followUpContainer_->setVisible(true);
  followUpContainer_->update();
}

void GameControls::showSecondLevelFollowUps(const QString& mainEvent, const QString& firstFollowUp) {
  Q_UNUSED(mainEvent);

  // Clear existing follow-up buttons and rebuild
  hideFollowUpButtons();

  QStringList actions = getSecondLevelFollowUps(currentMainEvent_, firstFollowUp);
  if (actions.isEmpty()) {
    // No second-level actions; finalize with the first follow-up
    emit gameEventMarked(currentMainEvent_, firstFollowUp);
    clearActiveMainButton();
    currentMainEvent_.clear();
    currentFirstFollowUp_.clear();
    followUpStage_ = FollowUpStage::None;
    return;
  }

  for (const QString& action : actions) {
    auto* button = new QPushButton(action, followUpContainer_);
    Style::setSize(button, "md");
    Style::setVariant(button, "gameControlFollowUp");
    button->setFocusPolicy(Qt::NoFocus);
    button->setMinimumHeight(40);

    connect(button, &QPushButton::clicked, this, [this, button]() { flashButtonBorder(button); });
    connect(button, &QPushButton::clicked, this, &GameControls::onFollowUpButtonClicked);

    followUpLayout_->addWidget(button);
    followUpButtons_.append(button);
  }

  followUpStage_ = FollowUpStage::SecondLevel;
  followUpContainer_->setVisible(true);
  followUpContainer_->update();
}

void GameControls::hideFollowUpButtons() {
  // Remove and delete all follow-up buttons
  for (auto* button : followUpButtons_) {
    // Clean up any flash timers for this button
    if (flashTimers_.contains(button)) {
      QTimer* timer = flashTimers_.take(button);
      if (timer) {
        timer->stop();
        timer->deleteLater();
      }
    }
    followUpLayout_->removeWidget(button);
    button->deleteLater();
  }
  followUpButtons_.clear();
  followUpContainer_->setVisible(false);
}

void GameControls::flashButtonBorder(QPushButton* button) {
  if (!button) return;

  QTimer* timer = flashTimers_.value(button, nullptr);
  if (!timer) {
    timer = new QTimer(button);
    timer->setSingleShot(true);
    flashTimers_.insert(button, timer);

    connect(timer, &QTimer::timeout, this, [button]() {
      if (!button) return;
      button->setProperty("flash", false);
      button->style()->unpolish(button);
      button->style()->polish(button);
      button->update();
    });
  }

  button->setProperty("flash", true);
  button->style()->unpolish(button);
  button->style()->polish(button);
  button->update();

  timer->start(150);
}

File: AVA_V02_py/components/game_controls.py
"""Main/follow-up game event controls used for tagging."""

from qt_compat import QtCore, QtGui, QtWidgets, Signal
from styles import set_size, set_state, set_variant


class GameControls(QtWidgets.QWidget):
    main_event_pressed = Signal(str)
    game_event_marked = Signal(str, str)

    def __init__(self, parent: QtWidgets.QWidget | None = None) -> None:
        super().__init__(parent)
        self.setMinimumWidth(300)
        self.current_main_event = ""
        self.current_first_follow_up = ""
        self.follow_up_stage = "none"
        self.active_main_button: QtWidgets.QPushButton | None = None
        self.follow_up_buttons: list[QtWidgets.QPushButton] = []
        self._flash_timers: dict[QtWidgets.QPushButton, QtCore.QTimer] = {}
        self._build_ui()
        self._wire_signals()
        self._build_shortcuts()
        self.hide_follow_up_buttons()

    def _build_ui(self) -> None:
        root = QtWidgets.QVBoxLayout(self)
        root.setContentsMargins(0, 0, 0, 0)
        root.setSpacing(12)

        main_grid_widget = QtWidgets.QWidget(self)
        self.main_grid_layout = QtWidgets.QGridLayout(main_grid_widget)
        self.main_grid_layout.setContentsMargins(0, 0, 0, 0)
        self.main_grid_layout.setSpacing(4)

        self.hit16yd_button = QtWidgets.QPushButton("16-yd play", main_grid_widget)
        self.hit50yd_button = QtWidgets.QPushButton("50-yd play", main_grid_widget)
        self.hit75yd_button = QtWidgets.QPushButton("75-yd play", main_grid_widget)
        self.enter_d_button = QtWidgets.QPushButton("Circle Entry", main_grid_widget)
        self.shot_button = QtWidgets.QPushButton("Shot", main_grid_widget)
        self.goal_button = QtWidgets.QPushButton("Goal", main_grid_widget)
        self.pc_button = QtWidgets.QPushButton("PC", main_grid_widget)
        self.ps_button = QtWidgets.QPushButton("PS", main_grid_widget)
        self.card_button = QtWidgets.QPushButton("Card", main_grid_widget)

        self.main_buttons = [
            self.hit16yd_button,
            self.hit50yd_button,
            self.hit75yd_button,
            self.enter_d_button,
            self.shot_button,
            self.goal_button,
            self.pc_button,
            self.ps_button,
            self.card_button,
        ]

        for button in self.main_buttons:
            set_size(button, "md")
            set_variant(button, "gameControl")
            button.setFocusPolicy(QtCore.Qt.FocusPolicy.StrongFocus)
            button.setMinimumHeight(40)

        self.main_grid_layout.addWidget(self.hit16yd_button, 0, 0)
        self.main_grid_layout.addWidget(self.hit50yd_button, 0, 1)
        self.main_grid_layout.addWidget(self.hit75yd_button, 0, 2)
        self.main_grid_layout.addWidget(self.enter_d_button, 1, 0)
        self.main_grid_layout.addWidget(self.shot_button, 1, 1)
        self.main_grid_layout.addWidget(self.goal_button, 1, 2)
        self.main_grid_layout.addWidget(self.pc_button, 2, 0)
        self.main_grid_layout.addWidget(self.ps_button, 2, 1)
        self.main_grid_layout.addWidget(self.card_button, 2, 2)

        self.follow_up_container = QtWidgets.QWidget(self)
        self.follow_up_layout = QtWidgets.QHBoxLayout(self.follow_up_container)
        self.follow_up_layout.setContentsMargins(0, 0, 0, 0)
        self.follow_up_layout.setSpacing(8)

        root.addWidget(main_grid_widget)
        root.addWidget(self.follow_up_container)
        root.addStretch(1)

    def _wire_signals(self) -> None:
        for button in self.main_buttons:
            button.clicked.connect(lambda _=False, b=button: self._on_main_button_clicked(b))

    def _build_shortcuts(self) -> None:
        def make_action(key_sequence: QtGui.QKeySequence, handler):
            action = QtGui.QAction(self)
            action.setShortcut(key_sequence)
            action.setShortcutContext(QtCore.Qt.ShortcutContext.ApplicationShortcut)
            action.triggered.connect(handler)
            self.addAction(action)
            return action

        mapping = {
            QtCore.Qt.Key.Key_Q: self.hit16yd_button,
            QtCore.Qt.Key.Key_W: self.hit50yd_button,
            QtCore.Qt.Key.Key_E: self.hit75yd_button,
            QtCore.Qt.Key.Key_A: self.enter_d_button,
            QtCore.Qt.Key.Key_S: self.shot_button,
            QtCore.Qt.Key.Key_D: self.goal_button,
            QtCore.Qt.Key.Key_Z: self.pc_button,
            QtCore.Qt.Key.Key_X: self.ps_button,
            QtCore.Qt.Key.Key_C: self.card_button,
        }

        for key, button in mapping.items():
            make_action(QtGui.QKeySequence(key), lambda b=button: self._click_if_available(b))

        for idx in range(1, 10):
            key = getattr(QtCore.Qt.Key, f"Key_{idx}")
            make_action(QtGui.QKeySequence(key), lambda i=idx - 1: self._trigger_follow_up_index(i))

        make_action(QtGui.QKeySequence(QtCore.Qt.Key.Key_Escape), self._cancel_follow_up)

    def _click_if_available(self, button: QtWidgets.QPushButton) -> None:
        if button.isVisible() and button.isEnabled():
            button.click()

    def _trigger_follow_up_index(self, index: int) -> None:
        if not self.follow_up_container.isVisible():
            return
        if index < 0 or index >= len(self.follow_up_buttons):
            return
        button = self.follow_up_buttons[index]
        if button.isVisible() and button.isEnabled():
            button.click()

    def _cancel_follow_up(self) -> None:
        if self.follow_up_stage == "none" or not self.current_main_event:
            return
        self.game_event_marked.emit(self.current_main_event, "")
        self.clear_active_main_button()
        self.hide_follow_up_buttons()
        self.current_main_event = ""
        self.current_first_follow_up = ""
        self.follow_up_stage = "none"

    def _set_active_main_button(self, button: QtWidgets.QPushButton | None) -> None:
        if self.active_main_button is button:
            return
        if self.active_main_button is not None:
            set_state(self.active_main_button, "activeMain", False)
        self.active_main_button = button
        if self.active_main_button is not None:
            set_state(self.active_main_button, "activeMain", True)

    def clear_active_main_button(self) -> None:
        if self.active_main_button is not None:
            set_state(self.active_main_button, "activeMain", False)
        self.active_main_button = None

    def _on_main_button_clicked(self, button: QtWidgets.QPushButton) -> None:
        self._flash_button_border(button)
        event_name = button.text()
        self.current_main_event = event_name
        self.current_first_follow_up = ""
        self.follow_up_stage = "none"
        self._set_active_main_button(button)
        self.main_event_pressed.emit(event_name)
        self.show_first_level_follow_ups(event_name)

    def _on_follow_up_button_clicked(self, button: QtWidgets.QPushButton) -> None:
        self._flash_button_border(button)
        follow_up_name = button.text()

        if self.follow_up_stage == "first":
            self.current_first_follow_up = follow_up_name
            second = self.get_second_level_follow_ups(self.current_main_event, self.current_first_follow_up)
            if not second:
                self.game_event_marked.emit(self.current_main_event, self.current_first_follow_up)
                self.clear_active_main_button()
                self.hide_follow_up_buttons()
                self.current_main_event = ""
                self.current_first_follow_up = ""
                self.follow_up_stage = "none"
                return

            self.show_second_level_follow_ups(self.current_main_event, self.current_first_follow_up)
            return

        if self.follow_up_stage == "second":
            combined = (
                follow_up_name
                if not self.current_first_follow_up
                else f"{self.current_first_follow_up} → {follow_up_name}"
            )
            self.game_event_marked.emit(self.current_main_event, combined)
            self.clear_active_main_button()
            self.hide_follow_up_buttons()
            self.current_main_event = ""
            self.current_first_follow_up = ""
            self.follow_up_stage = "none"
            return

        self.game_event_marked.emit(self.current_main_event, follow_up_name)

    def get_first_level_follow_ups(self, main_event: str) -> list[str]:
        if main_event == "Shot":
            return ["On target", "Off target"]
        if main_event == "Circle Entry":
            return ["Dribling", "Pass", "Deflection"]
        if main_event == "PC":
            return ["Direct shot", "Variant", "Ruined"]
        if main_event == "75-yd play":
            return ["Forward", "Sideways", "Back"]
        if main_event == "Card":
            return ["Green", "Yellow", "Red"]
        return []

    def get_second_level_follow_ups(self, main_event: str, first_follow_up: str) -> list[str]:
        if main_event == "Shot":
            if first_follow_up == "On target":
                return ["Goal", "Saved", "Post"]
            if first_follow_up == "Off target":
                return ["Closeby", "Not close"]
            return []

        if main_event == "PC" and first_follow_up == "Direct shot":
            return ["Hit", "Swept", "Dragflick"]

        if main_event == "Circle Entry":
            return ["Left", "Middle", "Right"]

        return []

    def show_first_level_follow_ups(self, main_event: str) -> None:
        self.hide_follow_up_buttons()
        actions = self.get_first_level_follow_ups(main_event)
        if not actions:
            self.game_event_marked.emit(main_event, "")
            self.clear_active_main_button()
            self.current_main_event = ""
            self.current_first_follow_up = ""
            self.follow_up_stage = "none"
            return

        for action in actions:
            button = QtWidgets.QPushButton(action, self.follow_up_container)
            set_size(button, "md")
            set_variant(button, "gameControlFollowUp")
            button.setFocusPolicy(QtCore.Qt.FocusPolicy.StrongFocus)
            button.setMinimumHeight(40)
            button.clicked.connect(lambda _=False, b=button: self._on_follow_up_button_clicked(b))
            self.follow_up_layout.addWidget(button)
            self.follow_up_buttons.append(button)

        self.follow_up_stage = "first"
        self.follow_up_container.setVisible(True)

    def show_second_level_follow_ups(self, main_event: str, first_follow_up: str) -> None:
        _ = main_event
        self.hide_follow_up_buttons()
        actions = self.get_second_level_follow_ups(self.current_main_event, first_follow_up)
        if not actions:
            self.game_event_marked.emit(self.current_main_event, first_follow_up)
            self.clear_active_main_button()
            self.current_main_event = ""
            self.current_first_follow_up = ""
            self.follow_up_stage = "none"
            return

        for action in actions:
            button = QtWidgets.QPushButton(action, self.follow_up_container)
            set_size(button, "md")
            set_variant(button, "gameControlFollowUp")
            button.setFocusPolicy(QtCore.Qt.FocusPolicy.StrongFocus)
            button.setMinimumHeight(40)
            button.clicked.connect(lambda _=False, b=button: self._on_follow_up_button_clicked(b))
            self.follow_up_layout.addWidget(button)
            self.follow_up_buttons.append(button)

        self.follow_up_stage = "second"
        self.follow_up_container.setVisible(True)

    def hide_follow_up_buttons(self) -> None:
        while self.follow_up_layout.count():
            item = self.follow_up_layout.takeAt(0)
            widget = item.widget()
            if widget is not None:
                widget.deleteLater()
        self.follow_up_buttons.clear()
        self.follow_up_container.setVisible(False)

    def _flash_button_border(self, button: QtWidgets.QPushButton) -> None:
        if button not in self._flash_timers:
            timer = QtCore.QTimer(button)
            timer.setSingleShot(True)
            timer.timeout.connect(lambda b=button: set_state(b, "flash", False))
            self._flash_timers[button] = timer

        set_state(button, "flash", True)
        self._flash_timers[button].start(150)
